Language Server Protocol (#1179)
JetBrains/intellij-sdk-docs: topics/reference_guide/custom_language_support/language_server_protocol.md

<!-- Copyright 2000-2023 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license. -->

> # Language Server Protocol

Reviewer: add "(LSP)" suffix in heading

> <link-summary>Language Server Protocol support in IntelliJ-based IDEs</link-summary>

Reviewer: add "(LSP)" suffix

The [Language Server Protocol](https://microsoft.github.io/language-server-protocol/) (LSP) is an open-standard protocol developed by Microsoft. It enables communication between development tools and Language Servers.
Language Servers can provide language-specific features such as code completion, documentation, and formatting, which is far easier than implementing language support from scratch.
It also reduces the need for constant maintenance and tracking of changes in relevant languages and tools, making it easier to bring consistent language support to various development environments.

> However, the canonical custom language support provided by IntelliJ Platform still offers a wider range of integration with IDE features than handling and presenting data provided by a Language Server.

Reviewer: link https://plugins.jetbrains.com/docs/intellij/custom-language-support.html for "canonic..."

- DataSpell
- RubyMine
- CLion
- Aqua
- DataGrip
- GoLand
- Rider

## Plugin Configuration

To fully utilize the LSP API in a third-party plugin based on the [](tools_gradle_intellij_plugin.md), it is recommended to upgrade the Gradle IntelliJ Plugin to the latest version available.
This plugin will attach the LSP API sources and code documentation to the project.

As LSP became available in the 2023.2 EAP7 of IntelliJ-based IDEs, the plugin must target IntelliJ IDEA Ultimate `2023.2` or later.

Example <path>build.gradle.kts</path> configuration:

```kotlin
plugins {
  // ...
  id("org.jetbrains.intellij") version "%gradle-intellij-plugin-version%"
}

intellij {
  version = "%ijPlatform%"
  type = "IU"
}
```

For projects based on the [](plugin_github_template.md), update the Gradle IntelliJ Plugin to the latest version, and amend the <path>gradle.properties</path> file as follows:

```
platformType = IU
platformVersion = %ijPlatform%
```

The <path>plugin.xml</path> configuration file needs to specify the dependency on the IntelliJ IDEA Ultimate module:

```xml
<idea-plugin>
   <!-- ... -->
   <depends>com.intellij.modules.ultimate</depends>
</idea-plugin>
```

> The LSP API sources are bundled in IntelliJ IDEA Ultimate and can be found within the <path>[IDE]/lib/src/src_lsp-openapi.zip</path> archive.

Reviewer: why does it matter? they're attached automatically to the project?

Author: As LSP is supported in other paid IDEs, only IU delivers sources one can use for development.

## Supported Features

> The initial LSP support within the IntelliJ Platform covers the following features:

Reviewer 2: I think now we may avoid the word `initial` and list the features as of 2023.3.

Since 2023.3:

- Intention actions ([textDocument/codeAction](https://microsoft.github.io/language-server-protocol/specification/#textDocument_codeAction))
- Code formatting ([textDocument/formatting](https://microsoft.github.io/language-server-protocol/specification/#textDocument_formatting))

Since 2023.3.1:

- Quick documentation ([textDocument/hover](https://microsoft.github.io/language-server-protocol/specification#textDocument_hover))

The first item could be about communication channel:

- Communication channels: STDIO (since 2023.2), sockets (since 2024.1)

- Errors and warnings highlighting ([textDocument/publishDiagnostics](https://microsoft.github.io/language-server-protocol/specification/#textDocument_publishDiagnostics))
- Quick fixes for errors and warnings ([textDocument/codeAction](https://microsoft.github.io/language-server-protocol/specification/#textDocument_codeAction))
- Code completion ([textDocument/completion](https://microsoft.github.io/language-server-protocol/specification/#textDocument_completion))
- Go to Declaration ([textDocument/definition](https://microsoft.github.io/language-server-protocol/specification/#textDocument_definition))

## Basic Implementation

A minimal LSP integration must implement `LspServerSupportProvider` along with a service descriptor and [register](plugin_extensions.md#declaring-extensions) it as a `com.intellij.platform.lsp.serverSupportProvider` [Extension Point (EP)](plugin_extension_points.md):

```kotlin
import com.intellij.platform.lsp.api.LspServerSupportProvider
import com.intellij.platform.lsp.api.ProjectWideLspServerDescriptor

> class FooLspServerSupportProvider : LspServerSupportProvider {

Reviewer: `internal class`

## Language Server Integration

Language Server is a separate process that analyzes source code and provides language-specific features to development tools.
When creating a plugin that utilizes LSP within the IDE, there are two possibilities for providing a Language Server to end-users:

- Bundle a [Language Server implementation](https://microsoft.github.io/language-server-protocol/implementors/servers/) binary as a resource delivered with a plugin.
- Provide a possibility for users to define the location of the Language Server binary in their environment.

The Prisma ORM plugin presents the first approach, which distributes the `prisma-language-server.js` script and uses a local Node.js interpreter to run it.

For more complex cases, the plugin may request to provide a detailed configuration with a dedicated [Settings](settings_guide.md) implementation.

To implement a minimal LSP plugin, perform the following steps:
- Within the `LspServerSupportProvider.fileOpened()` method, spin up the relevant LSP server descriptor, which can decide if the given file is supported by using the `LspServerDescriptor.isSupportedFile()` check method.

> - Tell how to start the server by implementing `LspServerDescriptor.createCommandLine()`.

Reviewer 2: I think this would be more natural in the `Basic Implementation` section, before the code sample that illustrates exactly these 2 points.

## Customization

> - To fine-tune or disable the implementation of LSP-based features, the plugins may override the corresponding properties of the `LspServerDescriptor` class.

Reviewer: the plugins -> plugins

  See the properties documentation for more details:
  - `lspGoToDefinitionSupport`
  - `lspCompletionSupport`
  - `lspDiagnosticsSupport`
  - `lspCodeActionsSupport`

>   - `lspCommandsSupport`

Reviewer 2: More items as of 2023.3:

  - `lspFormattingSupport`
  - `lspHoverSupport`

- To handle custom (undocumented) requests and notifications from the LSP server, override `LspServerDescriptor.createLsp4jClient`.
- To send custom (undocumented) requests and notifications to the LSP server, override `LspServerDescriptor.lsp4jServerClass` and implement the `LspClientNotification` and/or `LspRequest` classes.

See bundled LSP API source code and its documentation for more information.

## Troubleshooting

All the IDE and LSP server communication logs are passed to the IDE log file.

To include them for preview, add the following entry to the <control>Help | Diagnostic Tools | Debug Log Settings…</control> configuration dialog:

```
#com.intellij.platform.lsp
```

For more information, see the [](ide_infrastructure.md#logging) section.

## Limitations

- The current LSP API implementation assumes that the IDE <-> LSP server communication channel is `stdio`.
- The IDE doesn't send [workspace/didChangeWatchedFiles](https://microsoft.github.io/language-server-protocol/specification/#workspace_didChangeWatchedFiles) notifications to the server.

> 

Reviewer 2: I think we can remove this section.
`didChangeWatchedFiles` feature is available since didChangeWatchedFiles
`stdio` and `socket` transport can be mentioned in the `Supported Features` section

## Integration Overview

Integrating the Language Server Protocol (LSP) into a plugin for IntelliJ-based IDEs involves a trade-off between simple and fast language support and a complex custom language support plugin with IDE capabilities.

When considering the LSP-based approach, it is important to assess the following criteria for providing a Language Server to end users:

- OS dependency of the Language Server.
- Availability of the latest version online.
- Compatibility with breaking changes between versions.
- Feasibility of requesting the user to provide the Language Server binary path.

> If you encounter any issues or need assistance, please provide feedback by reaching out to us through the `#intellij-platform` channel in our [JetBrains Platform Slack](https://plugins.jetbrains.com/slack/) workspace or by submitting an issue in [YouTrack](https://youtrack.jetbrains.com/newIssue?project=IDEA&c=Subsystem%20Core.%20Platform%20API).

Reviewer: do not duplicate content from https://plugins.jetbrains.com/docs/intellij/getting-help.html#problems-with-code-support-issues

Reviewer: add section "sample plugins" with handpicked good samples
and link to https://plugins.jetbrains.com/intellij-platform-explorer/extensions?extensions=com.intellij.platform.lsp.serverSupportProvider for full list of existing LSP plugins

The Language Server can provide language-specific features such as code completion, documentation, and formatting, which is far easier than implementing language support from scratch.
It also reduces the need for constant maintenance and tracking of changes in relevant languages and tools, making it easier to bring consistent language support to various development environments.

However, we must remember that canonical custom language support provided by IntelliJ Platform still offers a wider range of integration with IDE features than handling and presenting data provided by a Language Server.
Therefore, the LSP approach shouldn't be considered as a replacement for the existing language API, but rather as an added value.

Starting with the 2023.2 release cycle, the LSP API is publicly available as part of the IntelliJ Platform in the following IDEs:
- IntelliJ IDEA Ultimate
- WebStorm
- PhpStorm
- PyCharm Professional
- DataSpell
- RubyMine
- CLion
- Aqua
- DataGrip
- GoLand
- Rider

> ## Plugin configuration

Reviewer 3: ```suggestion
## Plugin Configuration
```

This plugin will attach the LSP API sources and code documentation to your project.

As LSP became available in the 2023.2 EAP7 of IntelliJ-based IDEs, the plugin must target IntelliJ IDEA Ultimate `2023.2` or later.

Example <path>build.gradle.kts</path> configuration:

```kotlin
plugins {
  // ...
  id("org.jetbrains.intellij") version "1.15.0"
}

intellij {
  version = "232-EAP-SNAPSHOT"
  type = "IU"
}
```

For projects based on the [](plugin_github_template.md), update the Gradle IntelliJ Plugin to the latest version, and amend the <path>gradle.properties</path> file as follows:

```properties
platformType = IU
platformVersion = 232-EAP-SNAPSHOT
pluginSinceBuild = 232
pluginUntilBuild = 232.*
```

The <path>plugin.xml</path> configuration file needs to specify the dependency on the IntelliJ IDEA Ultimate module:

```xml
<idea-plugin>
   <!-- ... -->
   <depends>com.intellij.modules.platform</depends>
   <depends>com.intellij.modules.ultimate</depends>
</idea-plugin>
```

The LSP API sources are bundled in IntelliJ IDEA Ultimate and can be found within the <path>[IDE]/lib/src/src_lsp-openapi.zip</path> archive.

> ## Supported features

Reviewer 3: ```suggestion
## Supported Features
```

The initial LSP support within the IntelliJ Platform covers the following features:

- Errors/warnings highlighting ([textDocument/publishDiagnostics](https://microsoft.github.io/language-server-protocol/specification/#textDocument_publishDiagnostics))
- Quick-fixes for these errors/warnings ([textDocument/codeAction](https://microsoft.github.io/language-server-protocol/specification/#textDocument_codeAction))
- Code completion ([textDocument/completion](https://microsoft.github.io/language-server-protocol/specification/#textDocument_completion))
- Go to Declaration ([textDocument/definition](https://microsoft.github.io/language-server-protocol/specification/#textDocument_definition))
- More features will be announced soon.

> ## Basic implementation

Reviewer 3: ```suggestion
## Basic Implementation
```

> A minimal LSP integration must implement `LspServerSupportProvider` along with a service descriptor and register it as a `com.intellij.platform.lsp.serverSupportProvider` extension point:

Reviewer 3: "service descriptor" -> `LspServerDescriptor`?
Please also add links to these classes and make sure other class names have links on first mention.

```kotlin
import com.intellij.platform.lsp.api.LspServerSupportProvider
import com.intellij.platform.lsp.api.ProjectWideLspServerDescriptor

class FooLspServerSupportProvider : LspServerSupportProvider {
  override fun fileOpened(project: Project, file: VirtualFile, serverStarter: LspServerStarter) {
    if (file.extension == "foo") {
      serverStarter.ensureServerStarted(FooLspServerDescriptor(project))
    }
  }
}

private class FooLspServerDescriptor(project: Project) : ProjectWideLspServerDescriptor(project, "Foo") {
  override fun isSupportedFile(file: VirtualFile) = file.extension == "foo"
  override fun createCommandLine() = GeneralCommandLine("foo", "--stdio")
}
```

> After providing the implementation of the LSP Server support provider extension point, register it in the <path>plugin.xml</path> file as follows:

Reviewer 3: I think there is no need for this paragraph and the following snippet.

Reviewer 3: Does it differ from a regular extension registration? We describe registering tens of extensions in the documentation, and it's enough to just mention what should be registered in what EP. There is no need for code snippets. Also, this one doesn't show registration, but also adding dependencies (so it's not consistent with the introducing text), what was mentioned on the page already.

Author: Removed.

```xml
<idea-plugin>
   <!-- ... -->
   <depends>com.intellij.modules.platform</depends>
   <depends>com.intellij.modules.ultimate</depends>
   <extensions defaultExtensionNs="com.intellij">
       <platform.lsp.serverSupportProvider implementation="FooLspServerSupportProvider"/>
   </extensions>
</idea-plugin>
```

As a reference, check out the [Prisma ORM](https://plugins.jetbrains.com/plugin/20686-prisma-orm) open-source plugin implementation: https://github.com/JetBrains/intellij-plugins/tree/master/prisma/src/org/intellij/prisma/ide/lsp

> ## Language Server integration

Reviewer 3: ```suggestion
## Language Server Integration
```

Language Server is a separate process that analyzes source code and provides language-specific features to development tools.
When creating a plugin that utilizes LSP within the IDE, there are two possibilities for providing a Language Server to end-users:

- Bundle a [Language Server implementation](https://microsoft.github.io/language-server-protocol/implementors/servers/) binary as a resource delivered with a plugin.
- Provide a possibility for users to define the location of the Language Server binary in their environment.

The Prisma ORM plugin presents the first approach, which distributes the `prisma-language-server.js` script and uses a local Node.js interpreter to run it.

> For more complex cases, you may request to provide a detailed configuration with a dedicated [Settings](settings_guide.md) implementation.

Reviewer 3: ```suggestion
For more complex cases, the plugin may request to provide a detailed configuration with a dedicated [Settings](settings_guide.md) implementation.
```

> To implement a minimal but fully functional plugin, perform the following steps:

Reviewer 3: ```suggestion
To implement a minimal LSP plugin, perform the following steps:
```

> To implement a minimal but fully functional plugin, perform the following steps:

Reviewer 3: I think we don't need this paragraph at all. It is already described in the _Basic Implementation_ section. It doesn't add anything new here. Maybe we should just be more descriptive in the previous section about the steps, but IMHO there is no need, as there are excellent Javadocs.

- Within the `LspServerSupportProvider.fileOpened()` method, spin up the relevant LSP server descriptor, which can decide if the given file is supported by using the `LspServerDescriptor.isSupportedFile()` check method.
- Tell how to start the server by implementing `LspServerDescriptor.createCommandLine()`.

## Customization

- To fine-tune or disable the implementation of LSP-based features, the plugins may override the corresponding properties of the `LspServerDescriptor` class.
  See the properties documentation for more details:
  - `lspGoToDefinitionSupport`
  - `lspCompletionSupport`
  - `lspDiagnosticsSupport`
  - `lspCodeActionsSupport`
  - `lspCommandsSupport`
- To handle custom (undocumented) requests and notifications from the LSP server, override `LspServerDescriptor.createLsp4jClient`.
- To send custom (undocumented) requests and notifications to the LSP server, override `LspServerDescriptor.lsp4jServerClass` and implement the `LspClientNotification` and/or `LspRequest` classes.

> See bundled LSP API source code and its code documentation for more information.

Reviewer 3: ```suggestion
See bundled LSP API source code and its documentation for more information.
```

## Testing

> Utility classes used for testing are not yet extracted for public use, but we'll provide suitable methods for testing LSP-based third-party plugins as soon as possible.

Reviewer 3: IMHO, we shouldn't mention it in the docs until it is ready.

## Troubleshooting

All the IDE and LSP server communication logs are passed to the IDE log file.

To include them for preview, add the following entry to the <control>Help | Diagnostic Tools | Debug Log Settings…</control> configuration dialog:

```
#com.intellij.platform.lsp
```

For more information, see the [](ide_infrastructure.md#logging) section.

## Limitations

- The current LSP API implementation assumes that the IDE <-> LSP server communication channel is `stdio`.
- The IDE doesn't send [workspace/didChangeWatchedFiles](https://microsoft.github.io/language-server-protocol/specification/#workspace_didChangeWatchedFiles) notifications to the server.

> ## Conclusion

Reviewer 3: IMHO this is not a _Conclusion_ section, but a regular one, which gives tips about what should be considered when implementing LSP support. I would rename it to _Considerations_ or something better.

Integrating the Language Server Protocol (LSP) into a plugin for IntelliJ-based IDEs involves a trade-off between simple and fast language support and a complex custom language support plugin with IDE capabilities.

When considering the LSP-based approach, it is important to assess the following criteria for providing a Language Server to end users:

- OS dependency of the Language Server.
- Availability of the latest version online.
- Compatibility with breaking changes between versions.
- Feasibility of requesting the user to provide the Language Server binary path.

> > The integration with the Language Server Protocol is created as an extension to the paid IntelliJ-based IDEs.

Reviewer 3: I would move it to the top. If someone targets only free IDEs, it will be a waste of time to read it all and get information that it is useless for them.
